beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java

```diff
@@ -223,12 +232,18 @@ private boolean putToQueue(@Nullable StreamElement element, @Nullable UnifiedCon
       throws InterruptedException {
 
     Pair<StreamElement, UnifiedContext> p = Pair.of(element, context);
-    while (!stopped.get()) {
-      if (queue.offer(p, 50, TimeUnit.MILLISECONDS)) {
-        return true;
+    if (stopped.get()) {
+      if (nackAllIncoming && context != null) {
+        context.nack();
+      }
+    } else {
+      while (!stopped.get()) {
+        if (queue.offer(p, 50, TimeUnit.MILLISECONDS)) {
+          return true;
+        }
       }
+      log.debug("{}: Finishing consumption due to source being stopped", name);
```

Comment:
If I understand this correctly:
- Stop can be called asynchronously
- If source is stopped, we want to nack current element

Nacking seems broken, in case `stop` signal is received inside `while` loop

Should the code look like this instead?
```java
      while (!stopped.get()) {
        if (queue.offer(p, 50, TimeUnit.MILLISECONDS)) {
          return true;
        }
      }
      if (nackAllIncoming && context != null) {
        return true;
        context.nack();
      }
      log.debug("{}: Finishing consumption due to source being stopped", name);
      return false;
```

Reply:
Hm, nice catch. That is probably the reason why `lastWrittenContext` is explicitly nacked in the `stop` method. I'll try this. :+1: